crates/lang/codegen/src/generator/events.rs

```diff
@@ -236,16 +235,19 @@ impl<'a> Events<'a> {
                         .ident()
                         .map(quote::ToTokens::into_token_stream)
                         .unwrap_or_else(|| quote_spanned!(span => #n));
-                    let field_type = topic_field.ty();
+                    let signature = syn::LitByteStr::new(
+                        format!("{}::{}::{}", contract_ident, event_ident,
+                            field_ident
+                        ).as_bytes(), span);
                     quote_spanned!(span =>
-                        .push_topic::<#field_type>(&self.#field_ident)
+                        .push_topic(&::ink_env::topics::PrefixedValue { value: &self.#field_ident, prefix: #signature })
```

Comment:
I know it sounds strange but in macro generated code always try to be as explicit as possible. This includes re-introducing the exact type of the `PrefixedValue<T>`. ;)